[WIP] Refactor and clean up code base

davidallendj opened the pull request:
Recreated PR from #43. 

This PR covers a range of changes to be made to Magellan's internal API and command-line interface, including but not limited to:

 - Changing the CLI flag names to be more consistent
 - Changing the internal API signatures to some of the functions more useful in other areas (like in tests)
 - Changing how files are saved to disk to use the "hive" partitioning format
 - Reimplementing some of the CollectAll function to use the CrawlBMC
 - Condensing the output to only include desired information
 - Remove old code using bmclib and that are unused
The list above is subject to change as needed and aims to not introduce any new features into the current code base. It will likely break API compatibility to some extent as things are being changed, cleaned up, and moved around.

davidallendj commented:
Quick note: all of the major tasks for this PR are completed, but I'm still doing some general cleaning up in some areas not covered by this PR's tasks.

davidallendj commented:
General clean up is done. Just need someone else to test just to confirm that everything works like expected.

alexlovelltroy commented on cmd/login.go:
```diff
@@ -7,9 +7,9 @@ import (
 
 	magellan "github.com/OpenCHAMI/magellan/internal"
-	"github.com/OpenCHAMI/magellan/internal/log"
+	"github.com/OpenCHAMI/magellan/internal/util"
```
Creating a util package is usually a sign that you've made some incorrect assumptions about composability/modularity.

https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

Is there a missing proper library here, perhaps called auth?  Could it be made more general to allow other auth methods to be used in the future?

davidallendj replied:
Nope. I believe that can be split into several other packages, but I just made it `util` for convenience.

alexlovelltroy reviewed (changes requested):
Please review your util package and see if it can be reorganized for better modularity.

https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

davidallendj commented:
> Please review your util package and see if it can be reorganized for better modularity.
> 
> https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

I get what the article is saying, but I only partially agree with the sentiment. I think you could argue that the example given about the `net/http` package having the `client.go` and `server.go` files is similar to having a `util` and then having files organized similarly. I think there is certainly a much stronger argument for this if the package is meant to be exported in `pkg`, but less so when using it in `internal`. However, I do think there are some changes that can be made here to make things a bit more organized here.

davidallendj commented:
Yep, I agree. I'll separate that out from the util stuff along with some other stuff that makes sense.

synackd commented:
Small thing, but I think [here](https://github.com/OpenCHAMI/magellan/pull/46/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R51-R54) it makes sense to have the default build target use `go build` and have the goreleaser build be done in a separate target. That way, running `make` doesn't assume that goreleaser is installed.

davidallendj commented:
> Small thing, but I think [here](https://github.com/OpenCHAMI/magellan/pull/46/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R51-R54) it makes sense to have the default build target use `go build` and have the goreleaser build be done in a separate target. That way, running `make` doesn't assume that goreleaser is installed.

Good catch, and I think that's a reasonable assumption to make.

synackd commented on cmd/update.go:
```diff
@@ -15,87 +17,87 @@ var (
-	updateCmd.Flags().StringVar(&protocol, "protocol", "https", "set the Redfish protocol")
-	updateCmd.Flags().StringVar(&firmwareUrl, "firmware-url", "", "set the path to the firmware")
-	updateCmd.Flags().StringVar(&firmwareVersion, "firmware-version", "", "set the version of firmware to be installed")
+	updateCmd.Flags().StringVar(&firmwareUrl, "firmware.url", "", "set the path to the firmware")
+	updateCmd.Flags().StringVar(&firmwareVersion, "firmware.version", "", "set the version of firmware to be installed")
```
Is there a reason we're changing the dashes to periods?

davidallendj replied:
The idea was to make the arg reflect the YAML structure of the config, but maybe that doesn't since?

synackd replied:
I see.

Personally, I think dashes are sufficient and are more consistent.

synackd reviewed (changes requested):
Just a few questions and fixes after a cursory scan of the changes.

davidallendj commented on README.md:
```diff
@@ -120,9 +130,7 @@ Flags:
-### Running the Tool
-
-There are three main commands to use with the tool: `scan`, `list`, and `collect`. To start a network scan for BMC nodes, use the `scan` command. If the port is not specified, `magellan` will probe ports 623 and 443 by default:
+To start a network scan for BMC nodes, use the `scan` command. If the port is not specified, `magellan` will probe ports 623 and 443 by default:
```
Yes, I think I need to remove the 623 since that's not included by default anymore.

alexlovelltroy reviewed (approved):
This code organization is clearer. I approve pending testing from @synackd

synackd commented:
@davidallendj Just FYI, the `spell` target [here](https://github.com/OpenCHAMI/magellan/pull/46/files/f7159c9b661796feb6741c2b62e839f0a42dd0dd..c0a6d3bb6ff328095cd3604b551c2df8b5419336#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R68) requires `github.com/golangci/misspell` to be installed (running `make`).

Also, running the target yields some errors in the README and CHANGELOG:

```
Executing target: spell
misspell -error -locale=US -w **.md
CHANGELOG.md:16:14: corrected "refelct" to "reflect"
README.md:173:301: corrected "accessbile" to "accessible"
README.md:231:300: corrected "retrived" to "retrieved"
make: *** [Makefile:68: spell] Error 2
```

davidallendj commented:
> @davidallendj Just FYI, the `spell` target [here](https://github.com/OpenCHAMI/magellan/pull/46/files/f7159c9b661796feb6741c2b62e839f0a42dd0dd..c0a6d3bb6ff328095cd3604b551c2df8b5419336#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R68) requires `github.com/golangci/misspell` to be installed (running `make`).
> 
> Also, running the target yields some errors in the README and CHANGELOG:
> 
> ```
> Executing target: spell
> misspell -error -locale=US -w **.md
> CHANGELOG.md:16:14: corrected "refelct" to "reflect"
> README.md:173:301: corrected "accessbile" to "accessible"
> README.md:231:300: corrected "retrived" to "retrieved"
> make: *** [Makefile:68: spell] Error 2
> ```

I've honestly never noticed this target, but it's nice to know that it works. I'm going to have to start using it.

synackd commented:
Yeah it's neat. Not sure we should include it in the default make target since it requires that go package to be installed. We could probably just document in the readme that it's required for default builds instead.

What do you think?

davidallendj commented:
Yeah, I think I'd prefer to not include it in the default target, but at least have it as an option if it is installed. I guess I can add that to the README somewhere.

synackd commented:
Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:

```make
spell: ## misspell
        $(call print-target)
        $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
```

synackd commented:
Also, do we want to change [this](https://github.com/OpenCHAMI/magellan/pull/46/files/f7159c9b661796feb6741c2b62e839f0a42dd0dd..c0a6d3bb6ff328095cd3604b551c2df8b5419336#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R59-R63) to `container` like we did for the configurator?

synackd commented:
Getting errors during the `lint` target:

```
Executing target: lint
golangci-lint run --fix
cmd/collect.go:93:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
        viper.BindPFlag("collect.driver", collectCmd.Flags().Lookup("driver"))
                       ^
cmd/collect.go:94:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
        viper.BindPFlag("collect.host", collectCmd.Flags().Lookup("host"))
                       ^
cmd/collect.go:95:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
        viper.BindPFlag("collect.port", collectCmd.Flags().Lookup("port"))
                       ^
cmd/collect.go:102:18: Error return value of `viper.BindPFlags` is not checked (errcheck)
        viper.BindPFlags(collectCmd.Flags())
                        ^
cmd/root.go:56:12: Error return value of `cmd.Help` is not checked (errcheck)
                        cmd.Help()
                                ^
cmd/root.go:86:18: Error return value of `viper.BindPFlags` is not checked (errcheck)
        viper.BindPFlags(rootCmd.Flags())
                        ^
cmd/update.go:15:2: var `port` is unused (unused)
        port             int
        ^
tests/api_test.go:46:27: Error return value of `magellan.CollectInventory` is not checked (errcheck)
        magellan.CollectInventory(&results, &magellan.CollectParams{})
                                 ^
tests/compatibility_test.go:63:20: Error return value of `client.MakeRequest` is not checked (errcheck)
        client.MakeRequest(nil, url, http.MethodGet, body, headers)
                          ^
tests/compatibility_test.go:29:13: printf: fmt.Sprintf format %s has arg host of wrong type *string (govet)
                url     = fmt.Sprintf("%s/redfish/v1", host)
                          ^
tests/compatibility_test.go:58:13: printf: fmt.Sprintf format %s has arg host of wrong type *string (govet)
                url     = fmt.Sprintf("%s/redfish/v1", host)
                          ^
internal/util/error.go:15:3: ineffectual assignment to i (ineffassign)
                i += 1
                ^
pkg/client/smd.go:57:39: SA5011: possible nil pointer dereference (staticcheck)
        fmt.Printf("%v (%v)\n%s\n", url, res.Status, string(body))
                                             ^
pkg/client/smd.go:58:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
        if res != nil {
           ^
make: *** [Makefile:73: lint] Error 1
```

synackd commented on cmd/crawl.go:
```diff
@@ -17,10 +17,9 @@ import (
-		"store information about the scan into cache after completion. To do so, use the 'collect'" +
-		"command instead\n" +
-		"\n" +
+	Long: "Crawl a single BMC for inventory information. This command does NOT store information\n" +
+		"store information about the scan into cache after completion. To do so, use the 'collect'\n" +
+		"command instead\n\n" +
```
"store information" appears twice:

```
# ./magellan crawl --help
Crawl a single BMC for inventory information. This command does NOT store information
store information about the scan into cache after completion. To do so, use the 'collect'
command instead
```

synackd replied:
This looks good with the latest commit.

synackd reviewed (changes requested):
Just one more thing for now...

davidallendj commented:
> Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> 
> ```makefile
> spell: ## misspell
>         $(call print-target)
>         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> ```

Are we not assuming that for `goreleaser` and `golangci-lint`?

davidallendj commented:
> Getting errors during the `lint` target:
> 
> ```
> Executing target: lint
> golangci-lint run --fix
> cmd/collect.go:93:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
>         viper.BindPFlag("collect.driver", collectCmd.Flags().Lookup("driver"))
>                        ^
> cmd/collect.go:94:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
>         viper.BindPFlag("collect.host", collectCmd.Flags().Lookup("host"))
>                        ^
> cmd/collect.go:95:17: Error return value of `viper.BindPFlag` is not checked (errcheck)
>         viper.BindPFlag("collect.port", collectCmd.Flags().Lookup("port"))
>                        ^
> cmd/collect.go:102:18: Error return value of `viper.BindPFlags` is not checked (errcheck)
>         viper.BindPFlags(collectCmd.Flags())
>                         ^
> cmd/root.go:56:12: Error return value of `cmd.Help` is not checked (errcheck)
>                         cmd.Help()
>                                 ^
> cmd/root.go:86:18: Error return value of `viper.BindPFlags` is not checked (errcheck)
>         viper.BindPFlags(rootCmd.Flags())
>                         ^
> cmd/update.go:15:2: var `port` is unused (unused)
>         port             int
>         ^
> tests/api_test.go:46:27: Error return value of `magellan.CollectInventory` is not checked (errcheck)
>         magellan.CollectInventory(&results, &magellan.CollectParams{})
>                                  ^
> tests/compatibility_test.go:63:20: Error return value of `client.MakeRequest` is not checked (errcheck)
>         client.MakeRequest(nil, url, http.MethodGet, body, headers)
>                           ^
> tests/compatibility_test.go:29:13: printf: fmt.Sprintf format %s has arg host of wrong type *string (govet)
>                 url     = fmt.Sprintf("%s/redfish/v1", host)
>                           ^
> tests/compatibility_test.go:58:13: printf: fmt.Sprintf format %s has arg host of wrong type *string (govet)
>                 url     = fmt.Sprintf("%s/redfish/v1", host)
>                           ^
> internal/util/error.go:15:3: ineffectual assignment to i (ineffassign)
>                 i += 1
>                 ^
> pkg/client/smd.go:57:39: SA5011: possible nil pointer dereference (staticcheck)
>         fmt.Printf("%v (%v)\n%s\n", url, res.Status, string(body))
>                                              ^
> pkg/client/smd.go:58:5: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
>         if res != nil {
>            ^
> make: *** [Makefile:73: lint] Error 1
> ```

Yeah, I saw this earlier when I asked @bcfriesen to build on this branch. The fix for this is in the works.

synackd commented:
> > Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> > ```makefile
> > spell: ## misspell
> >         $(call print-target)
> >         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> > ```
> 
> Are we not assuming that for `goreleaser` and `golangci-lint`?

I think that would probably be the safest option. What say you?

davidallendj commented:
> > > Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> > > ```makefile
> > > spell: ## misspell
> > >         $(call print-target)
> > >         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> > > ```
> > 
> > 
> > Are we not assuming that for `goreleaser` and `golangci-lint`?
> 
> I think that would probably be the safest option. What say you?

Maybe I don't understand how `go`'s install command work, but I thought that installing would put the binary at `GOPATH/bin` and `GOPATH` is automatically set when installing go? I figured if we're just calling `goreleaser` we can just call `misspell` the same way.

synackd commented:
> > > > Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> > > > ```makefile
> > > > spell: ## misspell
> > > >         $(call print-target)
> > > >         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> > > > ```
> > > 
> > > 
> > > Are we not assuming that for `goreleaser` and `golangci-lint`?
> > 
> > 
> > I think that would probably be the safest option. What say you?
> 
> Maybe I don't understand how `go`'s install command work, but I thought that installing would put the binary at `GOPATH/bin` and `GOPATH` is automatically set when installing go? I figured if we're just calling `goreleaser` we can just call `misspell` the same way.

The way that `GOPATH` works is if the variable is not set, the `go` tool will use `~/go`. However, for tools installed with `go install` to be called as commands without using the binary's path, the `PATH` variable needs to have `$GOPATH/bin` in it, or `~/go/bin` if `GOPATH` is not set. My reasoning is that since adding either of these paths to `PATH` is typically not done automatically, it's probably best we don't assume they are in the Makefile and use the full path.

If we wanted to be able to allow customization of the path, we could move the `GOPATH` assignment to a Make variable, e.g:

```make
GOPATH ?= $(shell echo $${GOPATH:-~/go})
...
.PHONY: spell
spell: ## misspell
        $(call print-target)
        $(GOPATH)/bin/misspell -error -locale=US -w **.md
```

That way, it could be changed with `make GOPATH=... spell`

davidallendj commented:
> > > > > Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> > > > > ```makefile
> > > > > spell: ## misspell
> > > > >         $(call print-target)
> > > > >         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> > > > > ```
> > > > 
> > > > 
> > > > Are we not assuming that for `goreleaser` and `golangci-lint`?
> > > 
> > > 
> > > I think that would probably be the safest option. What say you?
> > 
> > 
> > Maybe I don't understand how `go`'s install command work, but I thought that installing would put the binary at `GOPATH/bin` and `GOPATH` is automatically set when installing go? I figured if we're just calling `goreleaser` we can just call `misspell` the same way.
> 
> The way that `GOPATH` works is if the variable is not set, the `go` tool will use `~/go`. However, for tools installed with `go install` to be called as commands without using the binary's path, the `PATH` variable needs to have `$GOPATH/bin` in it, or `~/go/bin` if `GOPATH` is not set. My reasoning is that since adding either of these paths to `PATH` is typically not done automatically, it's probably best we don't assume they are in the Makefile and use the full path.
> 
> If we wanted to be able to allow customization of the path, we could move the `GOPATH` assignment to a Make variable, e.g:
> 
> ```makefile
> GOPATH ?= $(shell echo $${GOPATH:-~/go})
> ...
> .PHONY: spell
> spell: ## misspell
>         $(call print-target)
>         $(GOPATH)/bin/misspell -error -locale=US -w **.md
> ```
> 
> That way, it could be changed with `make GOPATH=... spell`

That seems reasonable. I'll assume that should be done for the other two as well?

synackd commented:
> > > > > > Maybe, but we shouldn't assume that `PATH` is set with `~/go/bin`, or whatever the GOPATH is set to. So if we do, we should probably ensure that it is using the one that is installed. You could do something like:
> > > > > > ```makefile
> > > > > > spell: ## misspell
> > > > > >         $(call print-target)
> > > > > >         $${GOPATH:-~/go}/bin/misspell -error -locale=US -w **.md
> > > > > > ```
> > > > > 
> > > > > 
> > > > > Are we not assuming that for `goreleaser` and `golangci-lint`?
> > > > 
> > > > 
> > > > I think that would probably be the safest option. What say you?
> > > 
> > > 
> > > Maybe I don't understand how `go`'s install command work, but I thought that installing would put the binary at `GOPATH/bin` and `GOPATH` is automatically set when installing go? I figured if we're just calling `goreleaser` we can just call `misspell` the same way.
> > 
> > 
> > The way that `GOPATH` works is if the variable is not set, the `go` tool will use `~/go`. However, for tools installed with `go install` to be called as commands without using the binary's path, the `PATH` variable needs to have `$GOPATH/bin` in it, or `~/go/bin` if `GOPATH` is not set. My reasoning is that since adding either of these paths to `PATH` is typically not done automatically, it's probably best we don't assume they are in the Makefile and use the full path.
> > If we wanted to be able to allow customization of the path, we could move the `GOPATH` assignment to a Make variable, e.g:
> > ```makefile
> > GOPATH ?= $(shell echo $${GOPATH:-~/go})
> > ...
> > .PHONY: spell
> > spell: ## misspell
> >         $(call print-target)
> >         $(GOPATH)/bin/misspell -error -locale=US -w **.md
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That way, it could be changed with `make GOPATH=... spell`
> 
> That seems reasonable. I'll assume that should be done for the other two as well?

That makes sense to me.

synackd commented:
The Makefile GOPATH stuff looks good.

synackd commented:
They indeed are!

However, the unit tests are failing:

```
Executing target: test
go test -race -covermode=atomic -coverprofile=coverage.out -coverpkg=./... ./...
?       github.com/OpenCHAMI/magellan   [no test files]
?       github.com/OpenCHAMI/magellan/cmd       [no test files]
?       github.com/OpenCHAMI/magellan/internal  [no test files]
?       github.com/OpenCHAMI/magellan/internal/cache    [no test files]
?       github.com/OpenCHAMI/magellan/internal/cache/postgresql [no test files]
?       github.com/OpenCHAMI/magellan/pkg/client        [no test files]
?       github.com/OpenCHAMI/magellan/pkg/crawler       [no test files]
?       github.com/OpenCHAMI/magellan/internal/cache/sqlite     [no test files]
?       github.com/OpenCHAMI/magellan/internal/util     [no test files]
?       github.com/OpenCHAMI/magellan/pkg/auth  [no test files]
--- FAIL: TestScanAndCollect (0.00s)
    api_test.go:38: expected to find at least one BMC node, but found none
--- FAIL: TestRedfishV1Availability (0.00s)
    compatibility_test.go:36: failed to make request to BMC: failed to make request: Get "localhost/redfish/v1": unsupported protocol scheme ""
{"level":"error","error":"failed to make request: Get \"localhost/redfish/v1\": unsupported protocol scheme \"\"","time":"2024-08-13T18:06:00Z","message":"failed to make request"}
{"level":"error","error":"endpoint must starts with http or https","time":"2024-08-13T18:06:00Z","message":"failed to connect to BMC"}
--- FAIL: TestExpectedProperties (0.00s)
    compatibility_test.go:91: failed to crawl BMC: endpoint must starts with http or https
FAIL
coverage: 15.8% of statements in ./...
FAIL    github.com/OpenCHAMI/magellan/tests     0.023s
FAIL
```

davidallendj commented:
> They indeed are!
> 
> However, the unit tests are failing:
> 
> ```
> Executing target: test
> go test -race -covermode=atomic -coverprofile=coverage.out -coverpkg=./... ./...
> ?       github.com/OpenCHAMI/magellan   [no test files]
> ?       github.com/OpenCHAMI/magellan/cmd       [no test files]
> ?       github.com/OpenCHAMI/magellan/internal  [no test files]
> ?       github.com/OpenCHAMI/magellan/internal/cache    [no test files]
> ?       github.com/OpenCHAMI/magellan/internal/cache/postgresql [no test files]
> ?       github.com/OpenCHAMI/magellan/pkg/client        [no test files]
> ?       github.com/OpenCHAMI/magellan/pkg/crawler       [no test files]
> ?       github.com/OpenCHAMI/magellan/internal/cache/sqlite     [no test files]
> ?       github.com/OpenCHAMI/magellan/internal/util     [no test files]
> ?       github.com/OpenCHAMI/magellan/pkg/auth  [no test files]
> --- FAIL: TestScanAndCollect (0.00s)
>     api_test.go:38: expected to find at least one BMC node, but found none
> --- FAIL: TestRedfishV1Availability (0.00s)
>     compatibility_test.go:36: failed to make request to BMC: failed to make request: Get "localhost/redfish/v1": unsupported protocol scheme ""
> {"level":"error","error":"failed to make request: Get \"localhost/redfish/v1\": unsupported protocol scheme \"\"","time":"2024-08-13T18:06:00Z","message":"failed to make request"}
> {"level":"error","error":"endpoint must starts with http or https","time":"2024-08-13T18:06:00Z","message":"failed to connect to BMC"}
> --- FAIL: TestExpectedProperties (0.00s)
>     compatibility_test.go:91: failed to crawl BMC: endpoint must starts with http or https
> FAIL
> coverage: 15.8% of statements in ./...
> FAIL    github.com/OpenCHAMI/magellan/tests     0.023s
> FAIL
> ```

The tests are being worked on in another PR. #42

synackd commented:
Understood. In that case, the only thing left from my perspective is deciding whether to keep the flag format changes (`-` versus `.`) or not.

synackd commented on cmd/update.go:
```diff
@@ -1,101 +1,99 @@
+	updateCmd.Flags().StringVar(&username, "username", "", "Set the BMC user")
+	updateCmd.Flags().StringVar(&password, "password", "", "Set the BMC password")
+	updateCmd.Flags().StringVar(&transferProtocol, "transfer-protocol", "HTTP", "Set the transfer protocol")
+	updateCmd.Flags().StringVar(&firmwareUrl, "firmware.url", "", "Set the path to the firmware")
+	updateCmd.Flags().StringVar(&firmwareVersion, "firmware.version", "", "Set the version of firmware to be installed")
```
I think it would be better to switch these back to dashes, i.e. `firmware-url` and `firmware-version`. Mixing periods with dashes (e.g. we already have `transfer-protocol` above) could get confusing.

davidallendj replied:
I see I haven't done it in this repo, but in other repos I have flags that are meant to be represented as a object reflecting the config structure using the dot notation, such as `jwks.url` and `jwks.fetch-retries` vs `jwks-url` and `jwks-fetch-retries`. I think that this notation is less confusing when there are flags like `client.access-token` and if you're thinking about the same hierarchy as in the config. So something like `transfer.protocol` wouldn't make sense unless there is a `transfer` object being represented in the config followed by a nested `protocol` property. I had the `bmc.host` and `bmc.port` flags using this notation before removing the flags altogether.

Also a side note: I think I'm going to change `transfer-protocol` to `scheme` so it's similar to the other subcommands.

synackd replied:
That's fair, and I can see the reasoning for organizing options that way.

I just know that when I see the list of flags in the help menu and see `.` mixed in with `-`, I now have to remember where to put the `.` and where to put the `-` versus remembering the config words and separating them by only `-`. Like in the help output for `update`:

```
Usage:
  magellan update hosts... [flags]

Flags:
      --component string           Set the component to upgrade (BMC|BIOS)
      --firmware.url string        Set the path to the firmware
      --firmware.version string    Set the version of firmware to be installed
  -h, --help                       help for update
      --password string            Set the BMC password
      --status                     Get the status of the update
      --transfer-protocol string   Set the transfer protocol (default "HTTP")
      --username string            Set the BMC user
```

I see the `--firmware...` flags separated by `.` and it looks a bit out of place when my instinct would be to want to use `--firmware-url` and `--firmware-version`.

Also, I think it would be different if you had started with `--firmware.url` and `--firmware.version` in the first place, but you're changing it to that from the `-` version.

But perhaps that is a personal preference of mine based on what I see being convention in other projects. @alexlovelltroy, @njones-lanl, and @travisbcotton, what do you think?

synackd replied:
Commit [5b1e3e7](https://github.com/OpenCHAMI/magellan/pull/46/commits/5b1e3e732eb3c230d8a1ea6f98e699b375b9e922) changes the `.` back to `-` and we've decided to resolve this for now to keep things moving. If the three of you (or anyone else) mentioned above has anything to offer and we decide to change this, we can discuss it further.

For now, I'm resolving this discussion.

synackd commented on cmd/update.go:
```diff
@@ -1,101 +1,99 @@
+	checkBindFlagError(viper.BindPFlag("update.password", updateCmd.Flags().Lookup("password")))
+	checkBindFlagError(viper.BindPFlag("update.transfer-protocol", updateCmd.Flags().Lookup("transfer-protocol")))
+	checkBindFlagError(viper.BindPFlag("update.protocol", updateCmd.Flags().Lookup("protocol")))
+	checkBindFlagError(viper.BindPFlag("update.firmware.url", updateCmd.Flags().Lookup("firmware.url")))
+	checkBindFlagError(viper.BindPFlag("update.firmware.version", updateCmd.Flags().Lookup("firmware.version")))
```
See my comment on lines 84-85 above.

synackd replied:
Resolving with the resolution of the above comment.

synackd commented on cmd/collect.go:
```diff
@@ -30,82 +26,75 @@ var collectCmd = &cobra.Command{
-	viper.BindPFlag("collect.force-update", collectCmd.Flags().Lookup("force-update"))
-	viper.BindPFlag("collect.cacert", collectCmd.Flags().Lookup("secure-tls"))
-	viper.BindPFlags(collectCmd.Flags())
+	checkBindFlagError(viper.BindPFlag("collect.host", collectCmd.Flags().Lookup("collect.host")))
+	checkBindFlagError(viper.BindPFlag("collect.username", collectCmd.Flags().Lookup("collect.username")))
+	checkBindFlagError(viper.BindPFlag("collect.password", collectCmd.Flags().Lookup("collect.password")))
+	checkBindFlagError(viper.BindPFlag("collect.scheme", collectCmd.Flags().Lookup("collect.scheme")))
+	checkBindFlagError(viper.BindPFlag("collect.protocol", collectCmd.Flags().Lookup("collect.protocol")))
+	checkBindFlagError(viper.BindPFlag("collect.output", collectCmd.Flags().Lookup("collect.output")))
+	checkBindFlagError(viper.BindPFlag("collect.force-update", collectCmd.Flags().Lookup("collect.force-update")))
+	checkBindFlagError(viper.BindPFlag("collect.cacert", collectCmd.Flags().Lookup("collect.cacert")))
+	checkBindFlagError(viper.BindPFlags(collectCmd.Flags()))
```
I'm getting bind flag errors when running any subcommand with `--help`:
```
# ./magellan collect --help
{"level":"error","error":"flag for \"collect.host\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.username\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.password\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.scheme\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.protocol\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.output\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.force-update\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"collect.cacert\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"concurrency\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"timeout\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"verbose\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"debug\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"access-token\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"cache\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
{"level":"error","error":"flag for \"update.protocol\" is nil","time":"2024-08-13T18:32:42Z","message":"failed to bind flag"}
[... help message follows ...]
```

davidallendj replied:
I think that's failing because of the `Lookup()`. I think each arg should just be something like `host` instead of `collect.host`.

davidallendj replied:
Seems like the `root` commands aren't binding for some reason.

synackd replied:
These errors are no longer appearing with [7285492](https://github.com/OpenCHAMI/magellan/pull/46/commits/7285492815c41e39c47fcc50bb78ab9465392b88).

synackd reviewed (changes requested):
I made an updated review off of the latest changes for the discussion on flag format (`.` versus `-`). I also found some errors occurring with flags from the `collect` subcommand, so I'm requesting changes to fix those.

synackd commented on cmd/update.go:
```diff
@@ -88,8 +88,7 @@ func init() {
 	checkBindFlagError(viper.BindPFlag("update.password", updateCmd.Flags().Lookup("password")))
-	checkBindFlagError(viper.BindPFlag("update.transfer-protocol", updateCmd.Flags().Lookup("transfer-protocol")))
-	checkBindFlagError(viper.BindPFlag("update.protocol", updateCmd.Flags().Lookup("protocol")))
+	checkBindFlagError(viper.BindPFlag("update.scheme", updateCmd.Flags().Lookup("scheme")))
```
I was going to ask about the difference between `protocol` and `transfer-protocol`, so having a single `scheme` option makes more sense.

synackd commented on cmd/collect.go:
```diff
@@ -30,82 +26,75 @@ var collectCmd = &cobra.Command{
 
+		// URL sanitanization for host argument
+		host = strings.TrimSuffix(host, "/")
+		host = strings.ReplaceAll(host, "//", "/")
```
This will cause issues in the protocol section of the URL. I.e. it will replace `https://` with `https:/`, which is problematic.

davidallendj replied:
Should be fixed now.

synackd replied:
This was moved to the internal `url` package with better processing, so I'll mark this as resolved.

synackd reviewed (changes requested):
Getting:
```json
{"level":"error","error":"failed to make request: Post \"https:/foobar.openchami.cluster/hsm/v2/Inventory/RedfishEndpoints\": http: no Host in request URL","time":"2024-08-14T16:14:41Z","message":"failed to add Redfish endpoint"}
```
because of replacing `//` with `/`.

synackd reviewed (changes requested):
I found a small issue that is causing a 503 error during testing that is an easy fix. Upon testing with this fix, I got corrrect 401 and 409 responses when running `collect` on a populated SMD, but got a 500 when populating a fresh SMD. I'm not currently sure if this is with SMD or Magellan (I'm leaning towards the former), we'll have to dig deeper to find out.

davidallendj commented on internal/url/url.go:
```diff
@@ -0,0 +1,116 @@
+		return "", fmt.Errorf("failed to parse URI: %w", err)
+	}
+	// Remove any trailing slashes
+	parsedURI.Path = strings.TrimSuffix(uri, "/")
+	// Collapse any doubled slashes
+	parsedURI.Path = strings.ReplaceAll(uri, "//", "/")
```
Ah yeah, that was a mistake. Let me fix that.

davidallendj replied:
Should be fixed now.

synackd replied:
Looks like it works.

synackd replied:
On a side note related to the SMD thing, it looks like the node information gets added despite the 500 error because if I run the same command immediately after, I get a 409:
```json
{"level":"error","error":"returned status code 409 when adding endpoint","time":"2024-08-14T19:54:12Z","message":"failed to add Redfish endpoint"}
```

synackd replied:
Going to resolve this thread since the change was made and tested.

synackd commented:
LGTM
